src/server/serialization/journal_serializer.h

```diff
@@ -0,0 +1,298 @@
+    *key = facade::ToSV(span);
+    payload_consumed_ += size;
+  }
+
+  void Consume(journal::ValueType* value) {
+    Consume(std::get_if<journal::KeyType>(value));
+  }
+
+  void Consume(journal::ListType* value) {
+    uint32_t slice_size = PopSize();
+    absl::Span<std::string_view> slice;
+    {
+      auto byte_span = payload_buf_.InputBuffer().subspan(payload_consumed_);
+      size_t align = (8 - uint64_t(byte_span.data()) % 8) % 8;
+
+      auto* ptr = reinterpret_cast<std::string_view*>(byte_span.data() + align);
+
+      slice = absl::Span<std::string_view>{ptr, slice_size / sizeof(std::string_view)};
+      payload_consumed_ += slice_size + 8;
+    }
+
+    for (auto& sv : slice) {
+      new (&sv) std::string_view{};
+      Consume(&sv);
+    }
+
+    *value = slice;
+  }
+
+  template <typename... Ts> void Consume(std::tuple<Ts...>* tuple) {
+    std::apply([this](auto&... args) { (Consume(&args), ...); }, *tuple);
+  }
+
+  void Consume(journal::Payload* pl) {
+    std::visit([this](auto& value) { Consume(&value); }, *pl);
+  }
+
+  // TODO: pass buffer to allow in-place initializing in direct storage
+  journal::Entry ReadEntry() {
+    payload_buf_.ConsumeInput(payload_buf_.InputLen());
+    payload_sizes_.clear();
+    payload_consumed_ = 0;
+
+    journal::Entry e;
+    e.code = ReadOpcode();
+    e.payload = InitPayloadType(e.code);
+
+    Read(&e.payload);
+    std::reverse(payload_sizes_.begin(), payload_sizes_.end());
+    Consume(&e.payload);
+
+    return e;
+  }
+
+ private:
+  io::Source* source_;
+  base::IoBuf payload_buf_{128, std::align_val_t{8}};
+  std::vector<uint32_t> payload_sizes_;
+  size_t payload_consumed_ = 0;
+};
+
+struct JournalExecutor {
+  void Execute(const journal::PldEmpty, const journal::OpCode code) {
+  }
+
+  void Execute(const journal::PldKeyValue kv, const journal::OpCode code) {
+    CHECK(code == journal::OpCode::SET);
+    auto [key, val_v] = kv;
+    auto val = *std::get_if<std::string_view>(&val_v);
+
+    auto t = new Transaction{cmd_registry_->FindCmd("SET")};
+    t->InitByKeys(0, ArgSlice{&val, 1});
+
+    auto cb = [key = key, val = val](Transaction* t, EngineShard* shard) {
+      SetCmd sg(t->GetOpArgs(shard));
+      return sg.Set({}, key, val);
+    };
+    t->ScheduleSingleHop(std::move(cb));
+  }
+
+  void Execute(const journal::ListType list, const journal::OpCode code) {
+    CHECK(code == journal::OpCode::DEL);
+    auto t = new Transaction{cmd_registry_->FindCmd("DEL")};
+    t->InitByKeys(0, list);
+
+    auto cb = [](const Transaction* t, EngineShard* shard) {
+      ArgSlice args = t->ShardArgsInShard(shard->shard_id());
+      GenericFamily::OpDel(t->GetOpArgs(shard), args);
+      return OpStatus::OK;
+    };
+    t->ScheduleSingleHop(std::move(cb));
+  }
+
+  void Execute(const journal::PldKeyList kl, const journal::OpCode code) {
+    CHECK(code == journal::OpCode::LPUSH);
+    auto [key, vals] = kl;
```

Comment:
ideally, we want to reuse the execution code. Why not try running `Service::DispatchCommand` ?

Reply:
Because dispatch command requires the full command
Whats the point of journaling then, if we can just send the raw commands over the wire (possibly compressed)? (like Redis does for replication)

Reply:
we send parts (i.e. mset x val1 y val2 can be divided) but you still need to perform the hop. So what are you gaining by running a dedicated function on "set x val1" instead of running a DispatchCommand. Note, I am not arguing but trying to understand what's the advantage.